feat: implement conditional arbitrary description

Nortsova opened the pull request:
## Description

Depends on at least one of the creating arbitrary transaction issues being completed.

Displaying arbitrary transaction actions in the activity feed.

Action metadata:
Metadata
- Default: “Arbitrary transaction by {creator_username}”
- One contract interaction added: “Arbitrary transaction with method mint by {creator_username}”
- Multiple contract interactions added: “Arbitrary transaction with multiple methods by {creator_username}”

## Testing

Step 1. Enable Voting Reputation extension
Step 2. Create Custom transaction action ("Manage Colony" -> "Custom transactions")
Step 3. Before entering data, verify that action description has title "Arbitrary transaction by leela"
<img width="692" alt="image" src="https://github.com/user-attachments/assets/563f7d7f-3ed3-4c35-82fd-5f09e4a979b2" />


Step 4. Contract address `0xeF841fe1611ce41bFCf0265097EFaf50486F5111 `
ABI input: 
```
[{"type":"constructor","payable":false,"inputs":[{"type":"string","name":"_name"},{"type":"string","name":"_symbol"},{"type":"uint8","name":"_decimals"}]},{"type":"event","anonymous":false,"name":"Approval","inputs":[{"type":"address","name":"src","indexed":true},{"type":"address","name":"guy","indexed":true},{"type":"uint256","name":"wad","indexed":false}]},{"type":"event","anonymous":false,"name":"Burn","inputs":[{"type":"address","name":"guy","indexed":true},{"type":"uint256","name":"wad","indexed":false}]},{"type":"event","anonymous":false,"name":"LogSetAuthority","inputs":[{"type":"address","name":"authority","indexed":true}]},{"type":"event","anonymous":false,"name":"LogSetOwner","inputs":[{"type":"address","name":"owner","indexed":true}]},{"type":"event","anonymous":false,"name":"MetaTransactionExecuted","inputs":[{"type":"address","name":"user","indexed":false},{"type":"address","name":"relayerAddress","indexed":false},{"type":"bytes","name":"functionSignature","indexed":false}]},{"type":"event","anonymous":false,"name":"Mint","inputs":[{"type":"address","name":"guy","indexed":true},{"type":"uint256","name":"wad","indexed":false}]},{"type":"event","anonymous":false,"name":"Transfer","inputs":[{"type":"address","name":"src","indexed":true},{"type":"address","name":"dst","indexed":true},{"type":"uint256","name":"wad","indexed":false}]},{"type":"function","name":"DOMAIN_SEPARATOR","constant":true,"stateMutability":"view","payable":false,"inputs":[],"outputs":[{"type":"bytes32"}]},{"type":"function","name":"PERMIT_TYPEHASH","constant":true,"stateMutability":"view","payable":false,"inputs":[],"outputs":[{"type":"bytes32"}]},{"type":"function","name":"allowance","constant":true,"stateMutability":"view","payable":false,"inputs":[{"type":"address","name":"src"},{"type":"address","name":"guy"}],"outputs":[{"type":"uint256"}]},{"type":"function","name":"approve","constant":false,"payable":false,"inputs":[{"type":"address","name":"guy"},{"type":"uint256","name":"wad"}],"outputs":[{"type":"bool"}]},{"type":"function","name":"authority","constant":true,"stateMutability":"view","payable":false,"inputs":[],"outputs":[{"type":"address"}]},{"type":"function","name":"balanceOf","constant":true,"stateMutability":"view","payable":false,"inputs":[{"type":"address","name":"src"}],"outputs":[{"type":"uint256"}]},{"type":"function","name":"burn","constant":false,"payable":false,"inputs":[{"type":"uint256","name":"wad"}],"outputs":[]},{"type":"function","name":"burn","constant":false,"payable":false,"inputs":[{"type":"address","name":"guy"},{"type":"uint256","name":"wad"}],"outputs":[]},{"type":"function","name":"decimals","constant":true,"stateMutability":"view","payable":false,"inputs":[],"outputs":[{"type":"uint8"}]},{"type":"function","name":"executeMetaTransaction","constant":false,"stateMutability":"payable","payable":true,"inputs":[{"type":"address","name":"_user"},{"type":"bytes","name":"_payload"},{"type":"bytes32","name":"_sigR"},{"type":"bytes32","name":"_sigS"},{"type":"uint8","name":"_sigV"}],"outputs":[{"type":"bytes"}]},{"type":"function","name":"getMetatransactionNonce","constant":true,"stateMutability":"view","payable":false,"inputs":[{"type":"address","name":"_user"}],"outputs":[{"type":"uint256","name":"nonce"}]},{"type":"function","name":"locked","constant":true,"stateMutability":"view","payable":false,"inputs":[],"outputs":[{"type":"bool"}]},{"type":"function","name":"mint","constant":false,"payable":false,"inputs":[{"type":"address","name":"guy"},{"type":"uint256","name":"wad"}],"outputs":[]},{"type":"function","name":"mint","constant":false,"payable":false,"inputs":[{"type":"uint256","name":"wad"}],"outputs":[]},{"type":"function","name":"name","constant":true,"stateMutability":"view","payable":false,"inputs":[],"outputs":[{"type":"string"}]},{"type":"function","name":"nonces","constant":true,"stateMutability":"view","payable":false,"inputs":[{"type":"address","name":"_user"}],"outputs":[{"type":"uint256","name":"nonce"}]},{"type":"function","name":"owner","constant":true,"stateMutability":"view","payable":false,"inputs":[],"outputs":[{"type":"address"}]},{"type":"function","name":"permit","constant":false,"payable":false,"inputs":[{"type":"address","name":"owner"},{"type":"address","name":"spender"},{"type":"uint256","name":"value"},{"type":"uint256","name":"deadline"},{"type":"uint8","name":"v"},{"type":"bytes32","name":"r"},{"type":"bytes32","name":"s"}],"outputs":[]},{"type":"function","name":"setAuthority","constant":false,"payable":false,"inputs":[{"type":"address","name":"authority_"}],"outputs":[]},{"type":"function","name":"setOwner","constant":false,"payable":false,"inputs":[{"type":"address","name":"owner_"}],"outputs":[]},{"type":"function","name":"symbol","constant":true,"stateMutability":"view","payable":false,"inputs":[],"outputs":[{"type":"string"}]},{"type":"function","name":"totalSupply","constant":true,"stateMutability":"view","payable":false,"inputs":[],"outputs":[{"type":"uint256"}]},{"type":"function","name":"transfer","constant":false,"payable":false,"inputs":[{"type":"address","name":"dst"},{"type":"uint256","name":"wad"}],"outputs":[{"type":"bool"}]},{"type":"function","name":"transferFrom","constant":false,"payable":false,"inputs":[{"type":"address","name":"src"},{"type":"address","name":"dst"},{"type":"uint256","name":"wad"}],"outputs":[{"type":"bool"}]},{"type":"function","name":"unlock","constant":false,"payable":false,"inputs":[],"outputs":[]},{"type":"function","name":"verify","constant":true,"stateMutability":"view","payable":false,"inputs":[{"type":"address","name":"_user"},{"type":"uint256","name":"_nonce"},{"type":"uint256","name":"_chainId"},{"type":"bytes","name":"_payload"},{"type":"bytes32","name":"_sigR"},{"type":"bytes32","name":"_sigS"},{"type":"uint8","name":"_sigV"}],"outputs":[{"type":"bool"}]}]
```

Step 5. Add only one transaction with decision method: Permissions
Step 6. Verify that with one transaction description changed to "Arbitrary transaction with method mint by leela"
<img width="675" alt="image" src="https://github.com/user-attachments/assets/c939bfd4-ce41-4308-b8cd-d95699f0fb9a" />

Step 7. Add a second transaction and verify that the action description changed to "Arbitrary transaction with multiple methods by leela":
<img width="695" alt="image" src="https://github.com/user-attachments/assets/bebac811-ea78-469c-8d25-26596011d8f8" />

Step 8. Submit form and verify the title at the Completed Action screen should be: "Arbitrary transaction with multiple methods by leela":
<img width="694" alt="image" src="https://github.com/user-attachments/assets/7a3c26ae-e892-4ba7-ac78-7ee5149f4cd1" />

Step 9. Okay, lets create more Custom Transactions actions. We need to have 6, with one and multiple transactions per each decision method.

Step 10. Repeat Step 2 - 4. Add only one transaction with Decision method **Permissions**
Step 11. Repeat Step 2 - 4. Add one transaction with Decision method **Voting reputation**
Step 12. Fully support and fully oppose stakes
Step 13. Repeat Step 2 - 4. Add two or more transactions with Decision method **Voting reputation**
Step 14. Fully support and fully oppose stakes
Step 15. Enable MultiSig Extension.
Step 16. Give leela and amy Owner MultiSig roles.
Step 17. Repeat Step 2 - 4. Add one transaction with Decision method **Multi Sig**
Step 18. Repeat Step 2 - 4. Add two or more transactions with Decision method **Multi Sig**
Step 19. Now let's check formats. Activity feed:
<img width="1331" alt="image" src="https://github.com/user-attachments/assets/2fba27b8-5dd0-4a01-a6e7-315d6d9cc811" />
Step 20. Dashboard recent activity feed:
<img width="1371" alt="image" src="https://github.com/user-attachments/assets/6c626896-2c5f-46a9-9d63-6ee3dd49721e" />
Step 21. Stakes tab in User Hub
<img width="798" alt="image" src="https://github.com/user-attachments/assets/e2bee1f1-e41e-4f50-9506-78c5a17265ff" />

Step 22. Navigate to http://localhost:9091/planex/activity 
Step 23. Filter Actions by "Custom transactions", verify that actions filtered correctly:
<img width="1157" alt="image" src="https://github.com/user-attachments/assets/93df9701-9371-4f82-9098-4a6117eb2c76" />


(Resolves | Contributes to) #3538

jakubcolony reviewed (changes requested):
Nice start on this PR! 💯 The descriptions seem to be working flawlessly in most cases

I still had a transaction without ABIs from testing #4098 in the colony and it caused the whole page to crash:
<img width="522" alt="image" src="https://github.com/user-attachments/assets/30acd37a-7b72-440d-a1ec-a31e53e7d616" />

Here's a mutation I ran to simulate no ABIs:
```gql
mutation DeleteAbis {
  updateColonyActionMetadata(
    input: {id: "<TRANSACTION_HASH>", arbitraryTxAbis: null}
  ) {
    id
  }
}
```

The issue included a filter for custom transaction to the activity feed but it's not there at the moment. Feel free to handle it in this PR or open a separate issue
<img width="706" alt="image" src="https://github.com/user-attachments/assets/dd825a9d-bf3f-4a8d-8631-22d12a517d44" />


----

Empty state:
<img width="698" alt="image" src="https://github.com/user-attachments/assets/5bcde804-c83e-4e69-b925-1713e3e32157" />
<img width="957" alt="image" src="https://github.com/user-attachments/assets/d27763b1-0a31-4320-bf6a-3be724afe1e1" />

One transaction:
<img width="695" alt="image" src="https://github.com/user-attachments/assets/460eaada-d1e2-4d38-af7e-35836226075e" />

Multiple transactions:
<img width="697" alt="image" src="https://github.com/user-attachments/assets/80bec998-44cc-435e-86d3-3efb2156972a" />


Activity feed:

<img width="952" alt="image" src="https://github.com/user-attachments/assets/eaec2598-ba73-493b-b464-8ff9f22358b3" />

Userhub stakes:
<img width="681" alt="image" src="https://github.com/user-attachments/assets/822c3f14-1ae9-4c78-a79d-662639072431" />

Recent activity:
<img width="923" alt="image" src="https://github.com/user-attachments/assets/eb36e93e-3039-4447-89b3-aac934d90f88" />

Nortsova commented:
Hey, @jakubcolony ! Thanks for your review, I fixed "No ABIs" case: 
<img width="695" alt="image" src="https://github.com/user-attachments/assets/0c851402-f8dc-420d-94c8-e0e5e7e0a00a" />

A filter was also added 👍 
<img width="1142" alt="image" src="https://github.com/user-attachments/assets/ea40100f-0688-41bf-b7ee-d7fcf2974bf9" />

jakubcolony reviewed (dismissed):
Thank you for addressing the issues, everything looks great now 🌟 

<img width="933" alt="image" src="https://github.com/user-attachments/assets/f62eb814-10e8-41ae-8f86-4b2a206eff5d" />

<img width="1282" alt="image" src="https://github.com/user-attachments/assets/04cbc8ae-0bdc-4332-836d-f82be6156dcf" />

Action type filter ✅ 
<img width="1288" alt="image" src="https://github.com/user-attachments/assets/cb892405-b974-4cfb-9554-f91ca15a7f03" />

jakubcolony commented on src/components/common/ColonyActions/helpers/getActionTitleValues.ts:
```diff
@@ -199,9 +208,19 @@ const useGetActionTitleValues = ({
     actionType,
   });
+
+  if (actionType.includes(ColonyActionType.MakeArbitraryTransaction)) {
+    const arbitraryMessageValues =
+      getFormatValuesArbitraryTransactions(actionData);
+    messageValues.arbitraryTransactionsLength =
+      arbitraryMessageValues.arbitraryTransactionsLength;
+    messageValues.arbitraryMethod = arbitraryMessageValues.arbitraryMethod;
+  }
+
+  return messageValues;
```
Sorry I missed that the first time, I believe these should be added to `useMapColonyActionToExpectedFormat` hook

Nortsova replied:
you are right, thanks! fixed 👍

jakubcolony reviewed (commented):
Re-approving, thanks for the changes! 🙌 

<img width="696" alt="image" src="https://github.com/user-attachments/assets/bcf635c4-2ab3-4c52-a06e-f7d0c4ece26f" />

jakubcolony reviewed (approved):
Actually re-approving 🙂

iamsamgibbs reviewed (approved):
Great job with this @Nortsova it is all working very very nicely!

Filtering actions:
![Screenshot 2025-01-28 at 18 59 32](https://github.com/user-attachments/assets/b3f263e5-559c-450c-84bf-6d8ed05bcbe9)

Filtering correctly and all showing the correct descriptions.

UserHub transactions list:

![Screenshot 2025-01-28 at 18 59 06](https://github.com/user-attachments/assets/f44cc3bb-43f3-4958-bd23-efb04bea3ed1)

Fantastic work! Completely unrelated to this PR, but I did notice the motion pills are different in the userhub and the activity feed. One is completely wrong, and the other is wrong colour.

![Screenshot 2025-01-28 at 18 59 49](https://github.com/user-attachments/assets/892df9fa-49bc-4e7c-a155-b4c558b7894f)

rdig reviewed (approved):
Good to merge! Very nice!
